File: macapp/Sources/JornadaCore/Cedb.swift
import Foundation

/// Windows CE object-store database (CEDB) records, librapi2-compatible.
/// Mirrors jornada/cedb.py: a record travels over RAPI as an array of 16-byte
/// CEPROPVAL entries followed by the string and blob payloads they point to.
/// Pointer fields are byte offsets from the start of the buffer and every
/// payload starts 4-byte aligned — what librapi2's database.c writes and what
/// its pointer fix-up expects when reading.

/// CEVT_* value types — the low word of a CEPROPID.
public enum CedbType: Hashable, Sendable {
    case i2, i4, r8, bool, ui2, ui4, string, filetime, blob
    /// A type this codec does not know; its 8-byte value union is kept raw.
    case unknown(UInt16)

    public init(code: UInt16) {
        switch code {
        case 2: self = .i2
        case 3: self = .i4
        case 5: self = .r8
        case 11: self = .bool
        case 18: self = .ui2
        case 19: self = .ui4
        case 31: self = .string
        case 64: self = .filetime
        case 65: self = .blob
        default: self = .unknown(code)
        }
    }

    /// The CEVT_* value.
    public var code: UInt16 {
        switch self {
        case .i2: return 2
        case .i4: return 3
        case .r8: return 5
        case .bool: return 11
        case .ui2: return 18
        case .ui4: return 19
        case .string: return 31
        case .filetime: return 64
        case .blob: return 65
        case .unknown(let code): return code
        }
    }

    /// Short name as jornada.cedb's `kind_name` ("i2", "string", … or "0x007f").
    public var name: String {
        switch self {
        case .i2: return "i2"
        case .i4: return "i4"
        case .r8: return "r8"
        case .bool: return "bool"
        case .ui2: return "ui2"
        case .ui4: return "ui4"
        case .string: return "string"
        case .filetime: return "filetime"
        case .blob: return "blob"
        case .unknown(let code): return String(format: "0x%04x", code)
        }
    }
}

/// A property's decoded value; which case applies follows its `CedbType`.
public enum PropValue: Hashable, Sendable {
    case int(Int32)        // i2, i4
    case uint(UInt32)      // ui2, ui4
    case bool(Bool)
    case double(Double)    // r8
    case filetime(UInt64)  // 100 ns ticks since 1601-01-01
    case string(String)
    case blob(Data)
    case raw(Data)         // unknown type: the 8-byte CEVALUNION as received
    case missing           // CEDB_PROPNOTFOUND

    /// The zero/empty value of `kind` (what a CEDB_PROPDELETE entry carries).
    static func empty(for kind: CedbType) -> PropValue {
        switch kind {
        case .i2, .i4: return .int(0)
        case .ui2, .ui4: return .uint(0)
        case .bool: return .bool(false)
        case .r8: return .double(0)
        case .filetime: return .filetime(0)
        case .string: return .string("")
        case .blob: return .blob(Data())
        case .unknown: return .raw(Data(count: 8))
        }
    }
}

/// One property of a record: identifier, CEVT type, value and wFlags.
public struct PropVal: Hashable, Sendable {
    public let propId: UInt16
    public let kind: CedbType
    public let value: PropValue
    public let flags: UInt16

    public init(propId: UInt16, kind: CedbType, value: PropValue, flags: UInt16 = 0) {
        self.propId = propId
        self.kind = kind
        self.value = value
        self.flags = flags
    }

    /// The CEPROPID: identifier in the high word, CEVT type in the low word.
    public var cepropid: UInt32 { Cedb.propid(propId, kind) }
    public var isDeleted: Bool { flags & Cedb.propDelete != 0 }
    public var isMissing: Bool { flags & Cedb.propNotFound != 0 }

    public static func i2(_ propId: UInt16, _ value: Int16) -> PropVal {
        PropVal(propId: propId, kind: .i2, value: .int(Int32(value)))
    }

    public static func i4(_ propId: UInt16, _ value: Int32) -> PropVal {
        PropVal(propId: propId, kind: .i4, value: .int(value))
    }

    public static func ui2(_ propId: UInt16, _ value: UInt16) -> PropVal {
        PropVal(propId: propId, kind: .ui2, value: .uint(UInt32(value)))
    }

    public static func ui4(_ propId: UInt16, _ value: UInt32) -> PropVal {
        PropVal(propId: propId, kind: .ui4, value: .uint(value))
    }

    public static func bool(_ propId: UInt16, _ value: Bool) -> PropVal {
        PropVal(propId: propId, kind: .bool, value: .bool(value))
    }

    public static func r8(_ propId: UInt16, _ value: Double) -> PropVal {
        PropVal(propId: propId, kind: .r8, value: .double(value))
    }

    public static func filetime(_ propId: UInt16, _ ticks: UInt64) -> PropVal {
        PropVal(propId: propId, kind: .filetime, value: .filetime(ticks))
    }

    public static func string(_ propId: UInt16, _ value: String) -> PropVal {
        PropVal(propId: propId, kind: .string, value: .string(value))
    }

    public static func blob(_ propId: UInt16, _ value: Data) -> PropVal {
        PropVal(propId: propId, kind: .blob, value: .blob(value))
    }

    /// A property marked for removal on the next write (CEDB_PROPDELETE).
    public static func deleted(propId: UInt16, kind: CedbType) -> PropVal {
        PropVal(propId: propId, kind: kind, value: .empty(for: kind), flags: Cedb.propDelete)
    }
}

/// A database record: its object identifier and properties.
public struct Record: Hashable, Sendable, Identifiable {
    public let oid: UInt32
    public let props: [PropVal]
    public var id: UInt32 { oid }

    public init(oid: UInt32, props: [PropVal]) {
        self.oid = oid
        self.props = props
    }

    public func get(_ propId: UInt16) -> PropVal? { props.first { $0.propId == propId } }
    public func value(_ propId: UInt16) -> PropValue? { get(propId)?.value }

    public func string(_ propId: UInt16) -> String? {
        if case .string(let text)? = value(propId) { return text }
        return nil
    }

    /// Signed and unsigned integer properties (i2, i4, ui2, ui4).
    public func int(_ propId: UInt16) -> Int? {
        switch value(propId) {
        case .int(let number)?: return Int(number)
        case .uint(let number)?: return Int(number)
        default: return nil
        }
    }

    public func filetime(_ propId: UInt16) -> UInt64? {
        if case .filetime(let ticks)? = value(propId) { return ticks }
        return nil
    }

    public func blob(_ propId: UInt16) -> Data? {
        if case .blob(let bytes)? = value(propId) { return bytes }
        return nil
    }
}

/// A record buffer is malformed or a value does not fit its type.
public enum CedbError: Error, CustomStringConvertible, Equatable, Sendable {
    case valueMismatch(propId: UInt16, kind: CedbType)
    case unsupportedType(propId: UInt16, kind: CedbType)
    case tooManyProperties(Int)
    case truncated(count: Int, size: Int)
    case badOffset(what: String, offset: Int, size: Int, recordSize: Int)

    public var description: String {
        switch self {
        case .valueMismatch(let propId, let kind):
            return "property \(Cedb.hex(propId)) value does not fit type \(kind.name)"
        case .unsupportedType(let propId, let kind):
            return "cannot pack property \(Cedb.hex(propId)) of type \(kind.name)"
        case .tooManyProperties(let count):
            return "\(count) properties exceed the 16-bit record property count"
        case .truncated(let count, let size):
            return "\(count) properties need \(count * Cedb.propValSize) bytes, record has \(size)"
        case .badOffset(let what, let offset, let size, let recordSize):
            return "\(what) at offset \(offset) (+\(size)) is outside the \(recordSize)-byte record"
        }
    }
}

/// The CEPROPVAL record codec (jornada.cedb.pack_record / unpack_record).
public enum Cedb {
    public static let propValSize = 16
    /// CEPROPVAL.wFlags
    public static let propNotFound: UInt16 = 0x0100
    public static let propDelete: UInt16 = 0x0200
    /// CeOpenDatabase flag: reading a record advances the cursor.
    public static let autoincrement: UInt32 = 0x0000_0001
    /// CeReadRecordProps flag: the device may grow the caller's buffer.
    public static let allowRealloc: UInt32 = 0x0000_0001
    public static let maxSortOrder = 4
    public static let maxDatabaseNameLength = 32

    /// Round `size` up to the next multiple of four (librapi2's ALIGN).
    public static func align(_ size: Int) -> Int { (size + 3) & ~3 }

    /// Compose a CEPROPID from the 16-bit property identifier and CEVT type.
    public static func propid(_ propId: UInt16, _ kind: CedbType) -> UInt32 {
        (UInt32(propId) << 16) | UInt32(kind.code)
    }

    static func hex(_ propId: UInt16) -> String { String(format: "0x%04x", propId) }

    // MARK: - Packing

    /// Serialize properties into a CeWriteRecordProps buffer.
    public static func pack(_ props: [PropVal]) throws -> Data {
        let payloads = props.map(payload(of:))
        let offsets = payloadOffsets(sizes: payloads.map(\.count), from: align(props.count * propValSize))
        let header = try zip(props, offsets).reduce(into: Data()) { data, pair in
            data.append(try entry(pair.0, payloadOffset: pair.1))
        }
        return header + payloadSection(payloads)
    }

    /// The bytes a property appends after the CEPROPVAL array (empty for scalars).
    private static func payload(of prop: PropVal) -> Data {
        switch (prop.kind, prop.value) {
        case (.string, .string(let text)): return WireWriter.wstr(text)
        case (.blob, .blob(let bytes)): return bytes
        default: return Data()
        }
    }

    /// Byte offset of each payload from the buffer start (0 for an empty payload).
    private static func payloadOffsets(sizes: [Int], from start: Int) -> [Int] {
        var next = start
        return sizes.map { size in
            defer { next = align(next + size) }
            return size == 0 ? 0 : next
        }
    }

    /// Payloads back to back, each padded to a 4-byte boundary.
    private static func payloadSection(_ payloads: [Data]) -> Data {
        payloads.reduce(into: Data()) { section, payload in
            section.append(payload)
            section.append(Data(count: align(section.count) - section.count))
        }
    }

    /// One 16-byte CEPROPVAL: propid, wLenData (0), wFlags, value union.
    private static func entry(_ prop: PropVal, payloadOffset: Int) throws -> Data {
        var writer = WireWriter()
        writer.u32(prop.cepropid)
        writer.u16(0)
        writer.u16(prop.flags)
        writer.bytes(try packValue(prop, payloadOffset: payloadOffset))
        return writer.data
    }

    /// The 8-byte CEVALUNION as librapi2's PreparePropValForWriting emits it.
    private static func packValue(_ prop: PropVal, payloadOffset: Int) throws -> Data {
        let offset = UInt32(truncatingIfNeeded: payloadOffset)
        switch (prop.kind, prop.value) {
        case (.string, .string):
            return pair(offset, 0)
        case (.blob, .blob(let bytes)):
            return pair(UInt32(truncatingIfNeeded: bytes.count), offset)
        case (.i2, .int(let number)), (.i4, .int(let number)):
            return pair(UInt32(bitPattern: number), 0)
        case (.ui2, .uint(let number)), (.ui4, .uint(let number)):
            return pair(number, 0)
        case (.bool, .bool(let flag)):
            return pair(flag ? 1 : 0, 0)
        case (.filetime, .filetime(let ticks)):
            return pair(UInt32(truncatingIfNeeded: ticks), UInt32(truncatingIfNeeded: ticks >> 32))
        case (.r8, .double(let number)):
            let bits = number.bitPattern
            return pair(UInt32(truncatingIfNeeded: bits), UInt32(truncatingIfNeeded: bits >> 32))
        case (.unknown, _):
            throw CedbError.unsupportedType(propId: prop.propId, kind: prop.kind)
        default:
            throw CedbError.valueMismatch(propId: prop.propId, kind: prop.kind)
        }
    }

    private static func pair(_ low: UInt32, _ high: UInt32) -> Data {
        var writer = WireWriter()
        writer.u32(low)
        writer.u32(high)
        return writer.data
    }

    // MARK: - Unpacking

    /// Parse a CeReadRecordProps buffer holding `count` properties.
    public static func unpack(_ data: Data, count: Int) throws -> [PropVal] {
        let buffer = RecordBuffer(data)
        guard count >= 0, count * propValSize <= buffer.count else {
            throw CedbError.truncated(count: count, size: buffer.count)
        }
        return try (0..<count).map { index in try unpackEntry(buffer, at: index * propValSize) }
    }

    private static func unpackEntry(_ buffer: RecordBuffer, at start: Int) throws -> PropVal {
        let cepropid = buffer.u32(at: start)
        let flags = buffer.u16(at: start + 6)
        let propId = UInt16(truncatingIfNeeded: cepropid >> 16)
        let kind = CedbType(code: UInt16(truncatingIfNeeded: cepropid))
        guard flags & propNotFound == 0 else {
            return PropVal(propId: propId, kind: kind, value: .missing, flags: flags)
        }
        let value = try unpackValue(kind: kind, propId: propId, buffer: buffer, union: start + 8)
        return PropVal(propId: propId, kind: kind, value: value, flags: flags)
    }

    private static func unpackValue(kind: CedbType, propId: UInt16, buffer: RecordBuffer,
                                    union: Int) throws -> PropValue {
        switch kind {
        case .string:
            return .string(try readString(buffer, at: Int(buffer.u32(at: union))))
        case .blob:
            let count = Int(buffer.u32(at: union))
            let offset = Int(buffer.u32(at: union + 4))
            return .blob(try slice(buffer, offset: offset, size: count, what: "blob \(hex(propId))"))
        case .i2: return .int(Int32(Int16(bitPattern: buffer.u16(at: union))))
        case .ui2: return .uint(UInt32(buffer.u16(at: union)))
        case .i4: return .int(Int32(bitPattern: buffer.u32(at: union)))
        case .ui4: return .uint(buffer.u32(at: union))
        case .bool: return .bool(buffer.u32(at: union) != 0)
        case .filetime: return .filetime(buffer.u64(at: union))
        case .r8: return .double(Double(bitPattern: buffer.u64(at: union)))
        case .unknown: return .raw(buffer.bytes(at: union, size: 8))
        }
    }

    /// NUL-terminated UTF-16LE string at `offset` (stops at the buffer end).
    private static func readString(_ buffer: RecordBuffer, at offset: Int) throws -> String {
        guard offset >= 0, offset < buffer.count else {
            throw CedbError.badOffset(what: "string", offset: offset, size: 0, recordSize: buffer.count)
        }
        var units: [UInt16] = []
        var index = offset
        while index + 1 < buffer.count {
            let unit = buffer.u16(at: index)
            if unit == 0 { break }
            units.append(unit)
            index += 2
        }
        return String(decoding: units, as: UTF16.self)
    }

    private static func slice(_ buffer: RecordBuffer, offset: Int, size: Int, what: String) throws -> Data {
        guard offset >= 0, size >= 0, offset + size <= buffer.count else {
            throw CedbError.badOffset(what: what, offset: offset, size: size, recordSize: buffer.count)
        }
        return buffer.bytes(at: offset, size: size)
    }
}

/// Little-endian reads over a record buffer; offsets are relative to its start
/// and callers bounds-check them.
private struct RecordBuffer {
    private let data: Data

    init(_ data: Data) { self.data = data }

    var count: Int { data.count }

    func bytes(at offset: Int, size: Int) -> Data {
        let start = data.startIndex + offset
        return data.subdata(in: start..<(start + size))
    }

    func u16(at offset: Int) -> UInt16 {
        bytes(at: offset, size: 2).withUnsafeBytes { $0.loadUnaligned(as: UInt16.self).littleEndian }
    }

    func u32(at offset: Int) -> UInt32 {
        bytes(at: offset, size: 4).withUnsafeBytes { $0.loadUnaligned(as: UInt32.self).littleEndian }
    }

    func u64(at offset: Int) -> UInt64 {
        UInt64(u32(at: offset)) | (UInt64(u32(at: offset + 4)) << 32)
    }
}

extension FileTime {
    /// Ticks (100 ns) between 1601-01-01 and 1970-01-01.
    public static let epochDelta: UInt64 = 116_444_736_000_000_000

    /// 100 ns ticks since 1601 to Date; nil for zero (jornada.cedb.filetime_to_unix).
    public static func toDate(ticks: UInt64) -> Date? {
        guard ticks != 0 else { return nil }
        let seconds = ticks >= epochDelta
            ? Double(ticks - epochDelta) / 10_000_000
            : -Double(epochDelta - ticks) / 10_000_000
        return Date(timeIntervalSince1970: seconds)
    }

    /// Date to 100 ns ticks since 1601 (jornada.cedb.unix_to_filetime); clamps at 1601.
    public static func fromDate(_ date: Date) -> UInt64 {
        let ticks = (date.timeIntervalSince1970 * 10_000_000).rounded(.toNearestOrEven)
        return UInt64(clamping: Int64(ticks) + Int64(epochDelta))
    }
}

File: jornada/cedb.py
"""Windows CE object-store database (CEDB) records, librapi2-compatible.

A record travels over RAPI as an array of 16-byte ``CEPROPVAL`` structures
followed by the string and blob payloads they point to. On the wire the
pointer fields are **byte offsets from the start of the buffer** and every
payload starts 4-byte aligned — exactly what librapi2's ``database.c`` writes
and what its pointer fix-up expects when reading.

Everything here is a pure function over immutable values.
"""
from __future__ import annotations

import struct
from dataclasses import dataclass
from typing import Any, Dict, Optional, Tuple

from . import wire
from .constants import FILETIME_EPOCH_DELTA

# CEVT_* value types (low word of a CEPROPID)
CEVT_I2 = 2
CEVT_I4 = 3
CEVT_R8 = 5
CEVT_BOOL = 11
CEVT_UI2 = 18
CEVT_UI4 = 19
CEVT_LPWSTR = 31
CEVT_FILETIME = 64
CEVT_BLOB = 65

# CEPROPVAL.wFlags
CEDB_PROPNOTFOUND = 0x0100
CEDB_PROPDELETE = 0x0200

PROPVAL_SIZE = 16
_ALIGNMENT = 4

KIND_NAMES = {
    CEVT_I2: "i2", CEVT_I4: "i4", CEVT_R8: "r8", CEVT_BOOL: "bool",
    CEVT_UI2: "ui2", CEVT_UI4: "ui4", CEVT_LPWSTR: "string",
    CEVT_FILETIME: "filetime", CEVT_BLOB: "blob",
}


class CedbError(ValueError):
    """A record buffer is malformed or a value does not fit its type."""


def align(size: int) -> int:
    """Round ``size`` up to the next multiple of four (librapi2's ``ALIGN``)."""
    return (size + _ALIGNMENT - 1) & ~(_ALIGNMENT - 1)


def propid(prop_id: int, kind: int) -> int:
    """Compose a CEPROPID from the 16-bit property identifier and CEVT type."""
    return ((prop_id & 0xFFFF) << 16) | (kind & 0xFFFF)


@dataclass(frozen=True)
class PropVal:
    """One property of a record: identifier, CEVT type, value, wFlags."""

    prop_id: int
    kind: int
    value: Any
    flags: int = 0

    @property
    def propid(self) -> int:
        return propid(self.prop_id, self.kind)

    @property
    def kind_name(self) -> str:
        return KIND_NAMES.get(self.kind, f"0x{self.kind:04x}")

    # -- constructors ---------------------------------------------------------
    @classmethod
    def i2(cls, prop_id: int, value: int) -> "PropVal":
        return cls(prop_id, CEVT_I2, int(value))

    @classmethod
    def i4(cls, prop_id: int, value: int) -> "PropVal":
        return cls(prop_id, CEVT_I4, int(value))

    @classmethod
    def ui2(cls, prop_id: int, value: int) -> "PropVal":
        return cls(prop_id, CEVT_UI2, int(value))

    @classmethod
    def ui4(cls, prop_id: int, value: int) -> "PropVal":
        return cls(prop_id, CEVT_UI4, int(value))

    @classmethod
    def boolean(cls, prop_id: int, value: bool) -> "PropVal":
        return cls(prop_id, CEVT_BOOL, bool(value))

    @classmethod
    def r8(cls, prop_id: int, value: float) -> "PropVal":
        return cls(prop_id, CEVT_R8, float(value))

    @classmethod
    def filetime(cls, prop_id: int, ticks: int) -> "PropVal":
        return cls(prop_id, CEVT_FILETIME, int(ticks))

    @classmethod
    def string(cls, prop_id: int, value: str) -> "PropVal":
        return cls(prop_id, CEVT_LPWSTR, str(value))

    @classmethod
    def blob(cls, prop_id: int, value: bytes) -> "PropVal":
        return cls(prop_id, CEVT_BLOB, bytes(value))

    @classmethod
    def deleted(cls, prop_id: int, kind: int) -> "PropVal":
        """A property marked for removal on the next write (CEDB_PROPDELETE)."""
        empty: Any = "" if kind == CEVT_LPWSTR else b"" if kind == CEVT_BLOB else 0
        return cls(prop_id, kind, empty, CEDB_PROPDELETE)

    def to_json(self) -> Dict[str, Any]:
        value: Any = self.value
        if isinstance(value, (bytes, bytearray)):
            value = bytes(value).hex()   # blobs, and the raw union bytes of an unknown type
        return {"id": self.prop_id, "kind": self.kind_name, "value": value, "flags": self.flags}


@dataclass(frozen=True)
class Record:
    """A database record: its object identifier and properties."""

    oid: int
    props: Tuple[PropVal, ...]

    def get(self, prop_id: int) -> Optional[PropVal]:
        for prop in self.props:
            if prop.prop_id == prop_id:
                return prop
        return None

    def value(self, prop_id: int, default: Any = None) -> Any:
        prop = self.get(prop_id)
        return default if prop is None else prop.value

    def to_json(self) -> Dict[str, Any]:
        return {"oid": self.oid, "props": [p.to_json() for p in self.props]}


# -- FILETIME helpers ---------------------------------------------------------
def filetime_to_unix(ticks: int) -> Optional[float]:
    """100 ns ticks since 1601 → Unix seconds (None for zero)."""
    if ticks == 0:
        return None
    return (ticks - FILETIME_EPOCH_DELTA) / 10_000_000


def unix_to_filetime(seconds: float) -> int:
    return int(round(seconds * 10_000_000)) + FILETIME_EPOCH_DELTA


# -- packing ------------------------------------------------------------------
def _payload_size(prop: PropVal) -> int:
    if prop.kind == CEVT_LPWSTR:
        return len(wire.wstr(prop.value))
    if prop.kind == CEVT_BLOB:
        return len(prop.value)
    return 0


def _pack_value(prop: PropVal, payload_offset: int) -> bytes:
    """The 8-byte CEVALUNION as librapi2's PreparePropValForWriting emits it."""
    kind, value = prop.kind, prop.value
    if kind == CEVT_LPWSTR:
        return wire.u32(payload_offset) + wire.u32(0)
    if kind == CEVT_BLOB:
        return wire.u32(len(value)) + wire.u32(payload_offset)
    if kind == CEVT_I2:
        return struct.pack("<i", _check_range(prop, -0x8000, 0x7FFF)) + wire.u32(0)
    if kind == CEVT_I4:
        return struct.pack("<i", _check_range(prop, -0x80000000, 0x7FFFFFFF)) + wire.u32(0)
    if kind == CEVT_UI2:
        return struct.pack("<I", _check_range(prop, 0, 0xFFFF)) + wire.u32(0)
    if kind == CEVT_UI4:
        return struct.pack("<I", _check_range(prop, 0, 0xFFFFFFFF)) + wire.u32(0)
    if kind == CEVT_BOOL:
        return wire.u32(1 if value else 0) + wire.u32(0)
    if kind == CEVT_FILETIME:
        ticks = _check_range(prop, 0, 0xFFFFFFFFFFFFFFFF)
        return wire.u32(ticks & 0xFFFFFFFF) + wire.u32(ticks >> 32)
    if kind == CEVT_R8:
        return struct.pack("<d", float(value))
    raise CedbError(f"cannot pack property 0x{prop.prop_id:04x} of type 0x{kind:04x}")


def _check_range(prop: PropVal, low: int, high: int) -> int:
    value = int(prop.value)
    if not low <= value <= high:
        raise CedbError(f"property 0x{prop.prop_id:04x} value {value} out of range for {prop.kind_name}")
    return value


def pack_record(props: Tuple[PropVal, ...]) -> bytes:
    """Serialize properties into a CeWriteRecordProps buffer."""
    header = bytearray()
    payloads = bytearray()
    offset = align(len(props) * PROPVAL_SIZE)
    for prop in props:
        size = _payload_size(prop)
        header += wire.u32(prop.propid) + wire.u16(0) + wire.u16(prop.flags)
        header += _pack_value(prop, offset if size else 0)
        if prop.kind == CEVT_LPWSTR:
            payloads += wire.wstr(prop.value)
        elif prop.kind == CEVT_BLOB:
            payloads += bytes(prop.value)
        padded = align(offset + size)
        payloads += bytes(padded - offset - size)
        offset = padded
    return bytes(header) + bytes(payloads)


# -- unpacking ----------------------------------------------------------------
def _slice(data: bytes, offset: int, size: int, what: str) -> bytes:
    if offset < 0 or size < 0 or offset + size > len(data):
        raise CedbError(f"{what} at offset {offset} (+{size}) is outside the {len(data)}-byte record")
    return data[offset:offset + size]


def _read_wstr_at(data: bytes, offset: int) -> str:
    if offset < 0 or offset >= len(data):
        raise CedbError(f"string offset {offset} is outside the {len(data)}-byte record")
    raw = data[offset:]
    end = 0
    while end + 1 < len(raw) and raw[end:end + 2] != b"\x00\x00":
        end += 2
    return raw[:end].decode("utf-16-le", errors="replace")


def _unpack_value(kind: int, val: bytes, data: bytes, prop_id: int) -> Any:
    if kind == CEVT_LPWSTR:
        return _read_wstr_at(data, struct.unpack("<I", val[:4])[0])
    if kind == CEVT_BLOB:
        count, offset = struct.unpack("<II", val)
        return _slice(data, offset, count, f"blob 0x{prop_id:04x}")
    if kind == CEVT_I2:
        return struct.unpack("<h", val[:2])[0]
    if kind == CEVT_UI2:
        return struct.unpack("<H", val[:2])[0]
    if kind == CEVT_I4:
        return struct.unpack("<i", val[:4])[0]
    if kind in (CEVT_UI4, CEVT_BOOL):
        value = struct.unpack("<I", val[:4])[0]
        return bool(value) if kind == CEVT_BOOL else value
    if kind == CEVT_FILETIME:
        low, high = struct.unpack("<II", val)
        return (high << 32) | low
    if kind == CEVT_R8:
        return struct.unpack("<d", val)[0]
    return bytes(val)  # unknown type: keep the raw union bytes


def unpack_record(data: bytes, count: int) -> Tuple[PropVal, ...]:
    """Parse a CeReadRecordProps buffer holding ``count`` properties."""
    if count * PROPVAL_SIZE > len(data):
        raise CedbError(f"{count} properties need {count * PROPVAL_SIZE} bytes, record has {len(data)}")
    props = []
    for index in range(count):
        entry = data[index * PROPVAL_SIZE:(index + 1) * PROPVAL_SIZE]
        cepropid, _len_data, flags = struct.unpack("<IHH", entry[:8])
        prop_id, kind = cepropid >> 16, cepropid & 0xFFFF
        if flags & CEDB_PROPNOTFOUND:
            props.append(PropVal(prop_id, kind, None, flags))
            continue
        props.append(PropVal(prop_id, kind, _unpack_value(kind, entry[8:16], data, prop_id), flags))
    return tuple(props)
